apps/heft/src/plugins/CopyFilesPlugin.ts

```diff
@@ -0,0 +1,265 @@
+        });
+      });
+
+      build.hooks.postBuild.tap(PLUGIN_NAME, (postBuild: IPostBuildSubstage) => {
+        postBuild.hooks.run.tapPromise(HEFT_STAGE_TAP, async () => {
+          await this._runCopyFilesForHeftEvent(HeftEvent.postBuild, logger, heftConfiguration);
+        });
+      });
+    });
+  }
+
+  private async _runCopyFilesForHeftEvent(
+    heftEvent: HeftEvent,
+    logger: ScopedLogger,
+    heftConfiguration: HeftConfiguration
+  ): Promise<void> {
+    const eventActions: IHeftEventActions = await CoreConfigFiles.getConfigConfigFileEventActionsAsync(
+      logger.terminal,
+      heftConfiguration
+    );
+
+    const copyConfigurations: IExtendedSharedCopyConfiguration[] = [];
+    for (const copyFilesEventAction of eventActions.copyFiles.get(heftEvent) || []) {
+      copyConfigurations.push(...copyFilesEventAction.copyOperations);
+    }
+
+    await this.runCopyAsync({
+      buildFolder: heftConfiguration.buildFolder,
+      copyConfigurations,
+      logger
+    });
+  }
+
+  protected async runCopyAsync(options: ICopyFilesOptions): Promise<void> {
+    const { logger, buildFolder, copyConfigurations } = options;
+
+    const startTime: number = performance.now();
+    const copyDescriptors: ICopyFileDescriptor[] = await this._getCopyFileDescriptorsAsync(
+      buildFolder,
+      copyConfigurations
+    );
+
+    if (copyDescriptors.length === 0) {
+      // No need to run copy and print to console
+      return;
+    }
+
+    const [copyCount, hardlinkCount] = await this.copyFilesAsync(copyDescriptors);
+    const duration: number = performance.now() - startTime;
+    logger.terminal.writeLine(
+      `Copied ${copyCount} file${copyCount === 1 ? '' : 's'} and linked ${hardlinkCount} ` +
+        `file${hardlinkCount === 1 ? '' : 's'} in ${Math.round(duration)}ms`
+    );
+  }
+
+  protected async copyFilesAsync(copyDescriptors: ICopyFileDescriptor[]): Promise<[number, number]> {
+    if (copyDescriptors.length === 0) {
+      return [0, 0];
+    }
+
+    let copyCount: number = 0;
+    let hardlinkCount: number = 0;
+    await Async.forEachLimitAsync(
+      copyDescriptors,
+      MAX_PARALLELISM,
+      async (copyDescriptor: ICopyFileDescriptor) => {
+        if (copyDescriptor.hardlink) {
+          // Hardlink doesn't allow passing in overwrite param, so delete ourselves
+          try {
+            await FileSystem.deleteFileAsync(copyDescriptor.destinationFilePath, { throwIfNotExists: true });
+          } catch (e) {
+            if (!FileSystem.isFileDoesNotExistError(e)) {
+              throw e;
+            }
+            // Since the file doesn't exist, the parent folder may also not exist
+            await FileSystem.ensureFolderAsync(path.dirname(copyDescriptor.destinationFilePath));
+          }
+
+          await FileSystem.createHardLinkAsync({
+            linkTargetPath: copyDescriptor.sourceFilePath,
+            newLinkPath: copyDescriptor.destinationFilePath
+          });
```

Comment:
This feels fairly similar to [this code](https://github.com/microsoft/rushstack/blob/master/apps/heft/src/utilities/fileSystem/TypeScriptCachedFileSystem.ts#L154-L172)